### DISCUSSION TITLE
better SAFE_HEAP for O0/O1 using GLOBAL_BASE

### DISCUSSION THREAD

**Discussion by ghbgh** [emscripten-core/emscripten#21485] [Ideas]
```
    struct Test
    {
        int a;
        int b;
    };
    Test *p = nullptr;
    printf("%d\n", p->b);
```
This example does not throw exception when using -sSAFE_HEAP because &(p->b) 's address is not zero.
I found that there's a better check when using --low-memory-unused.
[https://github.com/WebAssembly/binaryen/blob/b3775b5e4e150439276ad3d34f1bb564b28e8ef5/src/passes/SafeHeap.cpp#L397](url)
It checks address < 1024 instead of = 0 when using GLOBAL_BASE >= 1024.
But for now emcc.py checks if optimizing level >= 2 before passing --low-memory-unused, so -O2 -sSAFE_HEAP -sGLOBAL_BASE=1024 can work on this case but -O0 can't.
[https://github.com/emscripten-core/emscripten/blob/e0bd10a0b61e08d1ed30966b6d8351ea06376fd4/tools/link.py#L349](url)
I wonder if it is better to simply remove this check?

**Comment by kripken** [comment]
IIUC you would like -O0/-O1 builds to also error on addresses < 1024?

The problem is that whether that range of addresses is valid or not depends on `--low-memory-unused`, which we apply as an optimization in -O2+. When low memory is unused, then accessing low memory is invalid, but not otherwise.

**Reply by ghbgh** [reply]
Yes, that's it.

**Reply by sbc100** [reply]
As Alon mentions address from 0 to 1024 are not invalid in -O0 builds since the stack memory starts at address zero in this mode.

We do this deliberately so that stack overflow errors result in invalid (negative) addresses.

One possibility would be to reserve for first 1024 bytes even in STACK_FIRST mode (maybe only in SAFE_HEAP + STACK_FIRST mode?).  I'm not sure its worth the complicity though.
